File: backend/workflow/src/main/java/eu/researchalps/api/UserFeedbackAdminApi.java
package eu.researchalps.api;

import eu.researchalps.api.util.ApiUtil;
import eu.researchalps.db.model.UserFeedback;
import eu.researchalps.db.model.UserFeedbackStatus;
import eu.researchalps.db.repository.UserFeedbackRepository;
import io.swagger.annotations.ApiOperation;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import static eu.researchalps.api.util.ApiConstants.PRODUCES_JSON;

/**
 * Copyright (C) by Data Publica, All Rights Reserved.
 */
@Controller
@RequestMapping("/admin/feedback")
public class UserFeedbackAdminApi {

    @Autowired
    private UserFeedbackRepository userFeedbackRepository;

    @ResponseBody
    @RequestMapping(method = RequestMethod.GET, produces = PRODUCES_JSON)
    @ApiOperation(value = "Get list of user feedbacks", notes = "Retrive the user feedbacks (active=true to retrieve only unprocessed feedbacks)")
    public Page<UserFeedback> getUserFeedbacks(@RequestParam(required = false, defaultValue = "1") int page,
                                               @RequestParam(required = false, defaultValue = "1000") int size,
                                               @RequestParam(required = false, defaultValue = "1") boolean active) {
        if (active)
            return userFeedbackRepository.findByStatus(UserFeedbackStatus.SUBMITTED, new PageRequest(page - 1, size, Sort.Direction.DESC, "creationDate"));
        else
            return userFeedbackRepository.findAll(new PageRequest(page - 1, size, Sort.Direction.DESC, "creationDate"));
    }

    @ResponseBody
    @RequestMapping(value = "/{id}", method = RequestMethod.GET, produces = PRODUCES_JSON)
    public UserFeedback getFeedback(@PathVariable String id) {
        return ApiUtil.fetchOrThrow(this.userFeedbackRepository, "userFeedback", id);
    }

    @ResponseBody
    @ApiOperation(value = "Mark a userFeedback as processed")
    @RequestMapping(value = "/{id}/markProcessed", method = RequestMethod.POST, produces = PRODUCES_JSON)
    public UserFeedback markFeedbackAsProcessed(@PathVariable String id) {
        UserFeedback userFeedback = ApiUtil.fetchOrThrow(this.userFeedbackRepository, "userFeedback", id);
        userFeedback.markProcessed();
        return userFeedbackRepository.save(userFeedback);
    }

    @ResponseBody
    @ApiOperation(value = "Create a userFeedback", notes = "Don't provide any id to create one")
    @RequestMapping(method = RequestMethod.PUT, produces = PRODUCES_JSON)
    public UserFeedback addUserFeedback(@RequestBody UserFeedback feedback) {
        return userFeedbackRepository.save(feedback);
    }
}

File: front/frontend-admin/app/api/UserFeedbackAdminApi.ts
/// <reference path="api.d.ts" />

/* tslint:disable:no-unused-variable member-ordering */

namespace api {
    'use strict';

    export class UserFeedbackAdminApi {
        protected basePath = '.';
        public defaultHeaders : any = {};

        static $inject: string[] = ['$http', '$httpParamSerializer'];

        constructor(protected $http: ng.IHttpService, protected $httpParamSerializer?: (d: any) => any, basePath?: string) {
            if (basePath) {
                this.basePath = basePath;
            }
        }

        private extendObj<T1,T2>(objA: T1, objB: T2) {
            for(let key in objB){
                if(objB.hasOwnProperty(key)){
                    objA[key] = objB[key];
                }
            }
            return <T1&T2>objA;
        }

        /**
         * Get list of user feedbacks
         * Retrive the user feedbacks (active=true to retrieve only unprocessed feedbacks)
         * @param page page
         * @param size size
         * @param active active
         */
        public getUserFeedbacks (page?: number, size?: number, active?: boolean, extraHttpRequestParams?: any ) : ng.IHttpPromise<PageOfUserFeedback> {
            const localVarPath = this.basePath + '/api/admin/feedback';

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            if (page !== undefined) {
                queryParameters['page'] = page;
            }

            if (size !== undefined) {
                queryParameters['size'] = size;
            }

            if (active !== undefined) {
                queryParameters['active'] = active;
            }

            let httpRequestParams: any = {
                method: 'GET',
                url: localVarPath,
                json: true,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * Create a userFeedback
         * Don&#39;t provide any id to create one
         * @param feedback feedback
         */
        public addUserFeedback (feedback: UserFeedback, extraHttpRequestParams?: any ) : ng.IHttpPromise<UserFeedback> {
            const localVarPath = this.basePath + '/api/admin/feedback';

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'feedback' is set
            if (!feedback) {
                throw new Error('Missing required parameter feedback when calling addUserFeedback');
            }
            let httpRequestParams: any = {
                method: 'PUT',
                url: localVarPath,
                json: true,
                data: feedback,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * getFeedback
         * 
         * @param id id
         */
        public getFeedback (id: string, extraHttpRequestParams?: any ) : ng.IHttpPromise<UserFeedback> {
            const localVarPath = this.basePath + '/api/admin/feedback/{id}'
                .replace('{' + 'id' + '}', String(id));

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'id' is set
            if (!id) {
                throw new Error('Missing required parameter id when calling getFeedback');
            }
            let httpRequestParams: any = {
                method: 'GET',
                url: localVarPath,
                json: true,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * Mark a userFeedback as processed
         * 
         * @param id id
         */
        public markFeedbackAsProcessed (id: string, extraHttpRequestParams?: any ) : ng.IHttpPromise<UserFeedback> {
            const localVarPath = this.basePath + '/api/admin/feedback/{id}/markProcessed'
                .replace('{' + 'id' + '}', String(id));

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'id' is set
            if (!id) {
                throw new Error('Missing required parameter id when calling markFeedbackAsProcessed');
            }
            let httpRequestParams: any = {
                method: 'POST',
                url: localVarPath,
                json: true,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
    }
}
